Dynamic vs. Static Mapping in ES

Currently, ES uses dynamic mapping to determine the type of fields. This caused us some issues with Solr, and appears to be happening in ES as well. For example, the `id_long` field in `mb11` (and `fullCaption` in `core18`) give the following errors because they're multi-valued:

```
2019-06-26 14:23:56,548 ERROR [pool-2-thread-31] index.IndexCollection$ESIndexerThread (IndexCollection.java:601) - Error sending bulk requests to Elasticsearch
com.fasterxml.jackson.core.JsonParseException: Duplicate field 'id_long'
 at [Source: (org.elasticsearch.common.bytes.BytesReference$MarkSupportingStreamInputWrapper); line: 1, column: 64]
        at com.fasterxml.jackson.core.json.JsonReadContext._checkDup(JsonReadContext.java:204) ~[jackson-core-2.9.9.jar:2.9.9]
        at com.fasterxml.jackson.core.json.JsonReadContext.setCurrentName(JsonReadContext.java:198) ~[jackson-core-2.9.9.jar:2.9.9]
        at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:734) ~[jackson-core-2.9.9.jar:2.9.9]
        at com.fasterxml.jackson.core.JsonGenerator.copyCurrentStructure(JsonGenerator.java:1877) ~[jackson-core-2.9.9.jar:2.9.9]
        at org.elasticsearch.common.xcontent.json.JsonXContentGenerator.copyCurrentStructure(JsonXContentGenerator.java:418) ~[elasticsearch-x-content-7.0.0.jar:7.0.0]
        at org.elasticsearch.common.xcontent.XContentBuilder.copyCurrentStructure(XContentBuilder.java:988) ~[elasticsearch-x-content-7.0.0.jar:7.0.0]
        at org.elasticsearch.client.RequestConverters.bulk(RequestConverters.java:239) ~[elasticsearch-rest-high-level-client-7.0.0.jar:7.0.0]
        at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1429) ~[elasticsearch-rest-high-level-client-7.0.0.jar:7.0.0]
        at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1403) ~[elasticsearch-rest-high-level-client-7.0.0.jar:7.0.0]
        at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1373) ~[elasticsearch-rest-high-level-client-7.0.0.jar:7.0.0]
        at org.elasticsearch.client.RestHighLevelClient.bulk(RestHighLevelClient.java:477) ~[elasticsearch-rest-high-level-client-7.0.0.jar:7.0.0]
        at io.anserini.index.IndexCollection$ESIndexerThread.sendBulkRequest(IndexCollection.java:597) [anserini-0.6.0-SNAPSHOT.jar:?]
        at io.anserini.index.IndexCollection$ESIndexerThread.run(IndexCollection.java:554) [anserini-0.6.0-SNAPSHOT.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_121]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_121]
        at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
        Suppressed: java.lang.IllegalStateException: Failed to close the XContentBuilder
                at org.elasticsearch.common.xcontent.XContentBuilder.close(XContentBuilder.java:1002) ~[elasticsearch-x-content-7.0.0.jar:7.0.0]
                at org.elasticsearch.client.RequestConverters.bulk(RequestConverters.java:238) ~[elasticsearch-rest-high-level-client-7.0.0.jar:7.0.0]
                at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1429) ~[elasticsearch-rest-high-level-client-7.0.0.jar:7.0.0]
                at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1403) ~[elasticsearch-rest-high-level-client-7.0.0.jar:7.0.0]
                at org.elasticsearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1373) ~[elasticsearch-rest-high-level-client-7.0.0.jar:7.0.0]
                at org.elasticsearch.client.RestHighLevelClient.bulk(RestHighLevelClient.java:477) ~[elasticsearch-rest-high-level-client-7.0.0.jar:7.0.0]
                at io.anserini.index.IndexCollection$ESIndexerThread.sendBulkRequest(IndexCollection.java:597) [anserini-0.6.0-SNAPSHOT.jar:?]
                at io.anserini.index.IndexCollection$ESIndexerThread.run(IndexCollection.java:554) [anserini-0.6.0-SNAPSHOT.jar:?]
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_121]
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_121]
                at java.lang.Thread.run(Thread.java:745) [?:1.8.0_121]
        Caused by: java.io.IOException: Unclosed object or array found
                at org.elasticsearch.common.xcontent.json.JsonXContentGenerator.close(JsonXContentGenerator.java:469) ~[elasticsearch-x-content-7.0.0.jar:7.0.0]
                at org.elasticsearch.common.xcontent.XContentBuilder.close(XContentBuilder.java:1000) ~[elasticsearch-x-content-7.0.0.jar:7.0.0]
                ... 10 more
```

There are a number of methods for defining the mapping (field types), so it should be investigated to see which option makes the most sense while also fixing the errors with multi-valued fields above. On one end, we have [explicit mappings](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping.html#_explicit_mappings) where the field types are defined beforehand. On the other end, we have [dynamic mappings](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping.html#_dynamic_mapping) (the default) which tries to infer field types. The middle ground seems to be [dynamic templates](https://www.elastic.co/guide/en/elasticsearch/reference/current/dynamic-templates.html) where we can define field properties based on data type, name pattern, etc.

Can you take a look, @charW?

--- a/docs/elastirini.md
+++ b/docs/elastirini.md
@@ -59,7 +59,41 @@ Once we have a local instance of Elasticsearch up and running, we can index usin
 First, let us create the index in Elasticsearch.
 
 ```
-curl --user elastic:changeme -XPUT -H 'Content-Type: application/json' 'localhost:9200/<index_name>' -d '{ "settings": { "index": { "refresh_interval": "60s", "similarity": { "default": { "type": "BM25", "k1": "0.9", "b": "0.4" }}}}}'
+
+curl --user elastic:changeme -XPUT -H 'Content-Type: application/json' 'localhost:9200/<index_name>' \
+    -d '{
+            "mappings": {
+                "dynamic_templates": [
+                    {
+                        "all_text": {
+                            "match_mapping_type": "string",
+                            "mapping": {
+                                "type": "text",
+                                "analyzer": "english"
+                            }
+                        }
+                    }
+                ],
+                "properties": {
+                    "published_date": {
+                        "type": "date",
+                        "format": "epoch_millis"
+                    }
+                }
+            },
+            "settings": {
+                "index": {
+                    "refresh_interval": "60s",
+                    "similarity": {
+                        "default": {
+                            "type": "BM25",
+                            "k1": "0.9",
+                            "b": "0.4"
+                        }
+                    }
+                }
+            }
+        }'
 ```
 
 Here, the username and password are those defaulted by `docker-elk`. You can change these if you like.

--- a/docs/elastirini.md
+++ b/docs/elastirini.md
@@ -75,6 +75,10 @@ curl --user elastic:changeme -XPUT -H 'Content-Type: application/json' 'localhos
                     }
                 ],
                 "properties": {
+                	"epoch": {
+                		"type": "date",
+                		"format": "epoch_second"
+                	},
                     "published_date": {
                         "type": "date",
                         "format": "epoch_millis"

--- a/docs/elastirini.md
+++ b/docs/elastirini.md
@@ -61,41 +61,41 @@ First, let us create the index in Elasticsearch.
 ```
 curl --user elastic:changeme -XPUT -H 'Content-Type: application/json' 'localhost:9200/<index_name>' \
     -d '{
-            "mappings": {
-                "dynamic_templates": [
-                    {
-                        "all_text": {
-                            "match_mapping_type": "string",
-                            "mapping": {
-                                "type": "text",
-                                "analyzer": "english"
-                            }
-                        }
-                    }
-                ],
-                "properties": {
-                	"epoch": {
-                		"type": "date",
-                		"format": "epoch_second"
-                	},
-                    "published_date": {
-                        "type": "date",
-                        "format": "epoch_millis"
-                    }
+          "mappings":{
+            "dynamic_templates":[
+              {
+                "all_text":{
+                  "match_mapping_type":"string",
+                  "mapping":{
+                    "type":"text",
+                    "analyzer":"english"
+                  }
                 }
-            },
-            "settings": {
-                "index": {
-                    "refresh_interval": "60s",
-                    "similarity": {
-                        "default": {
-                            "type": "BM25",
-                            "k1": "0.9",
-                            "b": "0.4"
-                        }
-                    }
-                }
+              }
+            ],
+            "properties":{
+              "epoch":{
+                "type":"date",
+                "format":"epoch_second"
+              },
+              "published_date":{
+                "type":"date",
+                "format":"epoch_millis"
+              }
             }
+          },
+          "settings":{
+            "index":{
+              "refresh_interval":"60s",
+              "similarity":{
+                "default":{
+                  "type":"BM25",
+                  "k1":"0.9",
+                  "b":"0.4"
+                }
+              }
+            }
+          }
         }'
 ```
 

--- a/docs/elastirini.md
+++ b/docs/elastirini.md
@@ -59,21 +59,15 @@ Once we have a local instance of Elasticsearch up and running, we can index usin
 First, let us create the index in Elasticsearch.
 
 ```
-curl --user elastic:changeme -XPUT "localhost:9200/yourindexname"
+curl --user elastic:changeme -XPUT -H 'Content-Type: application/json' 'localhost:9200/<index_name>' -d '{ "settings": { "index": { "refresh_interval": "60s", "similarity": { "default": { "type": "BM25", "k1": "0.9", "b": "0.4" }}}}}'
 ```
 
 Here, the username and password are those defaulted by `docker-elk`. You can change these if you like.
 
-You can further specify the settings associated with this index. For example, if you would like to change `index.refresh_interval` from the default 1 second to 60 seconds:
-
-```
-curl --user elastic:changeme -XPUT -H 'Content-Type: application/json' 'localhost:9200/yourindexname/_settings' -d '{ "index": { "refresh_interval": "60s"}}'
-```
-
 Now, we can start indexing through Elastirini. Here, instead of passing in `-index` (to index with Lucene directly) or `-solr` (to index with Solr), we pass in `-es`. For example, to index [robust04](https://github.com/castorini/anserini/blob/master/docs/regressions-robust04.md), we could run:
 
 ```
-sh target/appassembler/bin/IndexCollection -collection TrecCollection -generator JsoupGenerator -es -es.index yourindexname -threads 16 -input /absolute/path/to/disk45 -storePositions -storeDocvectors -storeRawDocs
+sh target/appassembler/bin/IndexCollection -collection TrecCollection -generator JsoupGenerator -es -es.index <index_name> -threads 16 -input /absolute/path/to/disk45 -storePositions -storeDocvectors -storeRawDocs
 ```
 
 There are also other `-es` parameters that you can specify as you see fit.

--- a/src/main/java/io/anserini/index/IndexCollection.java
+++ b/src/main/java/io/anserini/index/IndexCollection.java
@@ -33,22 +33,6 @@ import org.apache.commons.pool2.PooledObject;
 import org.apache.commons.pool2.impl.DefaultPooledObject;
 import org.apache.commons.pool2.impl.GenericObjectPool;
 import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
-import org.apache.lucene.analysis.CharArraySet;
-import org.apache.lucene.document.Document;
-import org.apache.lucene.index.ConcurrentMergeScheduler;
-import org.apache.lucene.index.DocValuesType;
-import org.apache.lucene.index.IndexWriter;
-import org.apache.lucene.index.IndexWriterConfig;
-import org.apache.lucene.index.IndexableField;
-import org.apache.lucene.index.Term;
-import org.apache.lucene.search.similarities.BM25Similarity;
-import org.apache.lucene.store.Directory;
-import org.apache.lucene.store.FSDirectory;
-import org.apache.solr.client.solrj.SolrClient;
-import org.apache.solr.client.solrj.impl.CloudSolrClient;
-import org.apache.solr.common.SolrInputDocument;
 import org.apache.http.HttpHost;
 import org.apache.http.auth.AuthScope;
 import org.apache.http.auth.UsernamePasswordCredentials;
@@ -56,15 +40,24 @@ import org.apache.http.client.CredentialsProvider;
 import org.apache.http.client.config.RequestConfig;
 import org.apache.http.impl.client.BasicCredentialsProvider;
 import org.apache.http.impl.nio.client.HttpAsyncClientBuilder;
-import org.elasticsearch.common.xcontent.XContentFactory;
-import org.elasticsearch.common.xcontent.XContentBuilder;
-import org.elasticsearch.client.RestHighLevelClient;
-import org.elasticsearch.client.RestClient;
-import org.elasticsearch.client.RestClientBuilder.HttpClientConfigCallback;
-import org.elasticsearch.client.RestClientBuilder.RequestConfigCallback;
-import org.elasticsearch.client.RequestOptions;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.analysis.CharArraySet;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.index.*;
+import org.apache.lucene.search.similarities.BM25Similarity;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FSDirectory;
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.impl.CloudSolrClient;
+import org.apache.solr.common.SolrInputDocument;
 import org.elasticsearch.action.bulk.BulkRequest;
 import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.client.RestHighLevelClient;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
 import org.kohsuke.args4j.*;
 
 import java.io.File;
@@ -72,16 +65,13 @@ import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Optional;
-import java.util.Set;
-import java.util.Iterator;
+import java.util.*;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public final class IndexCollection {
   private static final Logger LOG = LogManager.getLogger(IndexCollection.class);
@@ -179,7 +169,7 @@ public final class IndexCollection {
     @Option(name = "-solr.poolSize", metaVar = "[NUMBER]", usage = "the number of clients to keep in the pool")
     public int solrPoolSize = 16;
 
-    @Option(name="-es", forbids = {"-index", "-solr"}, usage = "boolean switch to determine if we should index through Elasticsearch")
+    @Option(name = "-es", forbids = {"-index", "-solr"}, usage = "boolean switch to determine if we should index through Elasticsearch")
     public boolean es = false;
 
     @Option(name = "-es.batch", usage = "the number of index requests in a bulk request sent to Elasticsearch")
@@ -277,14 +267,14 @@ public final class IndexCollection {
         @SuppressWarnings("unchecked")
         LuceneDocumentGenerator generator =
             (LuceneDocumentGenerator) generatorClass
-                  .getDeclaredConstructor(Args.class, Counters.class)
-                  .newInstance(args, counters);
+                .getDeclaredConstructor(Args.class, Counters.class)
+                .newInstance(args, counters);
 
         int cnt = 0;
 
         @SuppressWarnings("unchecked")
         FileSegment<SourceDocument> segment =
-                (FileSegment) collection.createFileSegment(inputFile);
+            (FileSegment) collection.createFileSegment(inputFile);
 
         for (Object document : segment) {
           SourceDocument d = (SourceDocument) document;
@@ -329,19 +319,19 @@ public final class IndexCollection {
         if (skipped > 0) {
           counters.skipped.addAndGet(skipped);
           LOG.warn(inputFile.getParent().getFileName().toString() + File.separator +
-                  inputFile.getFileName().toString() + ": " + skipped +
-                  " docs skipped.");
+              inputFile.getFileName().toString() + ": " + skipped +
+              " docs skipped.");
         }
 
         if (segment.getErrorStatus()) {
           counters.errors.incrementAndGet();
           LOG.error(inputFile.getParent().getFileName().toString() + File.separator +
-                  inputFile.getFileName().toString() + ": error iterating through segment.");
+              inputFile.getFileName().toString() + ": error iterating through segment.");
         }
 
         segment.close();
         LOG.info(inputFile.getParent().getFileName().toString() + File.separator +
-              inputFile.getFileName().toString() + ": " + cnt + " docs added.");
+            inputFile.getFileName().toString() + ": " + cnt + " docs added.");
         counters.indexed.addAndGet(cnt);
       } catch (Exception e) {
         LOG.error(Thread.currentThread().getName() + ": Unexpected Exception:", e);
@@ -365,15 +355,15 @@ public final class IndexCollection {
       try {
         @SuppressWarnings("unchecked")
         LuceneDocumentGenerator generator =
-                (LuceneDocumentGenerator) generatorClass
-                        .getDeclaredConstructor(Args.class, Counters.class)
-                        .newInstance(args, counters);
+            (LuceneDocumentGenerator) generatorClass
+                .getDeclaredConstructor(Args.class, Counters.class)
+                .newInstance(args, counters);
 
         int cnt = 0;
 
         @SuppressWarnings("unchecked")
         FileSegment<SourceDocument> segment =
-                (FileSegment) collection.createFileSegment(input);
+            (FileSegment) collection.createFileSegment(input);
 
         for (Object d : segment) {
           SourceDocument sourceDocument = (SourceDocument) d;
@@ -436,14 +426,14 @@ public final class IndexCollection {
         if (skipped > 0) {
           counters.skipped.addAndGet(skipped);
           LOG.warn(input.getParent().getFileName().toString() + File.separator +
-                  input.getFileName().toString() + ": " + skipped +
-                  " docs skipped.");
+              input.getFileName().toString() + ": " + skipped +
+              " docs skipped.");
         }
 
         if (segment.getErrorStatus()) {
           counters.errors.incrementAndGet();
           LOG.error(input.getParent().getFileName().toString() + File.separator +
-                  input.getFileName().toString() + ": error iterating through segment.");
+              input.getFileName().toString() + ": error iterating through segment.");
         }
 
         segment.close();
@@ -497,13 +487,13 @@ public final class IndexCollection {
 
         @SuppressWarnings("unchecked")
         LuceneDocumentGenerator generator =
-                (LuceneDocumentGenerator) generatorClass
-                        .getDeclaredConstructor(Args.class, Counters.class)
-                        .newInstance(args, counters);
+            (LuceneDocumentGenerator) generatorClass
+                .getDeclaredConstructor(Args.class, Counters.class)
+                .newInstance(args, counters);
 
         @SuppressWarnings("unchecked")
         FileSegment<SourceDocument> segment =
-                (FileSegment) collection.createFileSegment(input);
+            (FileSegment) collection.createFileSegment(input);
 
         int cnt = 0;
 
@@ -534,22 +524,43 @@ public final class IndexCollection {
             continue;
           }
 
-          XContentBuilder builder = XContentFactory.jsonBuilder();
-          builder.startObject();
-          for (IndexableField field : document.getFields()) {
-            if (field.name().equals(LuceneDocumentGenerator.FIELD_RAW) && !args.storeRawDocs) continue;
-            if (field.name().equals(LuceneDocumentGenerator.FIELD_BODY) && !args.storeTransformedDocs) continue;
+          // Get distinct field names
+          List<String> fields = document.getFields().stream().map(field -> field.name()).distinct().collect(Collectors.toList());
 
-            if (field.stringValue() != null) {
-              builder.field(field.name(), field.stringValue());
-            } else if (field.numericValue() != null) {
-              builder.field(field.name(), field.numericValue());
+          XContentBuilder builder = XContentFactory.jsonBuilder().startObject();
+
+          for (String field : fields) {
+
+            // Skip docValues fields
+            if (document.getField(field).fieldType().docValuesType() != DocValuesType.NONE) continue;
+
+            if (field.equals(LuceneDocumentGenerator.FIELD_RAW) && !args.storeRawDocs) continue;
+            if (field.equals(LuceneDocumentGenerator.FIELD_BODY) && !args.storeTransformedDocs) continue;
+
+            // Get field objects for current field name (could be multiple, such as WaPo's fullCaption)
+            IndexableField[] indexableFields = document.getFields(field);
+
+            if (field.equalsIgnoreCase("id") || indexableFields.length == 1) {
+              // Single value fields or "id" field
+              Object value = document.getField(field).stringValue() != null ? document.getField(field).stringValue() : document.getField(field).numericValue();
+              switch (field) {
+                case "published_date":
+                  builder.timeField(field, value);
+                default:
+                  builder.field(field, value);
+              }
+            } else {
+              // Multi-valued fields
+              Object[] values = Stream.of(indexableFields).map(f -> f.stringValue() != null ? f.stringValue() : f.numericValue()).toArray();
+              builder.array(field, values);
             }
+
           }
+
           builder.endObject();
-          
+
           String indexName = (args.esIndex != null) ? args.esIndex : input.getFileName().toString();
-          bulkRequest.add(new IndexRequest(indexName, "doc").id(sourceDocument.id()).source(builder));
+          bulkRequest.add(new IndexRequest(indexName).id(sourceDocument.id()).source(builder));
           if (bulkRequest.numberOfActions() == args.esBatch) {
             sendBulkRequest();
           }
@@ -565,14 +576,14 @@ public final class IndexCollection {
         if (skipped > 0) {
           counters.skipped.addAndGet(skipped);
           LOG.warn(input.getParent().getFileName().toString() + File.separator +
-                  input.getFileName().toString() + ": " + skipped +
-                  " docs skipped.");
+              input.getFileName().toString() + ": " + skipped +
+              " docs skipped.");
         }
 
         if (segment.getErrorStatus()) {
           counters.errors.incrementAndGet();
           LOG.error(input.getParent().getFileName().toString() + File.separator +
-                  input.getFileName().toString() + ": error iterating through segment.");
+              input.getFileName().toString() + ": error iterating through segment.");
         }
 
         segment.close();
@@ -734,17 +745,17 @@ public final class IndexCollection {
       final CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
       credentialsProvider.setCredentials(AuthScope.ANY, new UsernamePasswordCredentials(args.esUser, args.esPassword));
       return new RestHighLevelClient(
-        RestClient.builder(new HttpHost(args.esHostname, args.esPort, "http"))
-        .setHttpClientConfigCallback(
-                (HttpAsyncClientBuilder httpClientBuilder) -> 
-                        httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider))
-        .setRequestConfigCallback(
-                (RequestConfig.Builder requestConfigBuilder) -> requestConfigBuilder
-                .setConnectTimeout(args.esConnectTimeout)
-                .setSocketTimeout(args.esSocketTimeout))
+          RestClient.builder(new HttpHost(args.esHostname, args.esPort, "http"))
+              .setHttpClientConfigCallback(
+                  (HttpAsyncClientBuilder httpClientBuilder) ->
+                      httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider))
+              .setRequestConfigCallback(
+                  (RequestConfig.Builder requestConfigBuilder) -> requestConfigBuilder
+                      .setConnectTimeout(args.esConnectTimeout)
+                      .setSocketTimeout(args.esSocketTimeout))
       );
     }
-    
+
     @Override
     public PooledObject<RestHighLevelClient> wrap(RestHighLevelClient esClient) {
       return new DefaultPooledObject(esClient);
@@ -830,7 +841,7 @@ public final class IndexCollection {
       try {
         SolrClient client = solrPool.borrowObject();
         if (!args.dryRun) {
-            client.commit(args.solrIndex);
+          client.commit(args.solrIndex);
         }
         // Needed for orderly shutdown so the SolrClient executor does not delay main thread exit
         solrPool.returnObject(client);

--- a/src/main/java/io/anserini/index/IndexCollection.java
+++ b/src/main/java/io/anserini/index/IndexCollection.java
@@ -540,7 +540,7 @@ public final class IndexCollection {
               builder.field(field, value);
             } else {
               // Multi-valued fields
-              Object[] values = Stream.of(indexableFields).map(f -> f.stringValue() != null ? f.stringValue() : f.numericValue()).toArray();
+              Object[] values = Stream.of(indexableFields).map(f -> f.stringValue()).toArray();
               builder.array(field, values);
             }
           }

--- a/src/main/java/io/anserini/index/IndexCollection.java
+++ b/src/main/java/io/anserini/index/IndexCollection.java
@@ -37,9 +37,7 @@ import org.apache.http.HttpHost;
 import org.apache.http.auth.AuthScope;
 import org.apache.http.auth.UsernamePasswordCredentials;
 import org.apache.http.client.CredentialsProvider;
-import org.apache.http.client.config.RequestConfig;
 import org.apache.http.impl.client.BasicCredentialsProvider;
-import org.apache.http.impl.nio.client.HttpAsyncClientBuilder;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import org.apache.lucene.analysis.CharArraySet;
@@ -425,15 +423,12 @@ public final class IndexCollection {
         int skipped = segment.getSkippedCount();
         if (skipped > 0) {
           counters.skipped.addAndGet(skipped);
-          LOG.warn(input.getParent().getFileName().toString() + File.separator +
-              input.getFileName().toString() + ": " + skipped +
-              " docs skipped.");
+          LOG.warn(input.getParent().getFileName().toString() + File.separator + input.getFileName().toString() + ": " + skipped + " docs skipped.");
         }
 
         if (segment.getErrorStatus()) {
           counters.errors.incrementAndGet();
-          LOG.error(input.getParent().getFileName().toString() + File.separator +
-              input.getFileName().toString() + ": error iterating through segment.");
+          LOG.error(input.getParent().getFileName().toString() + File.separator + input.getFileName().toString() + ": error iterating through segment.");
         }
 
         segment.close();
@@ -467,7 +462,6 @@ public final class IndexCollection {
         }
       }
     }
-
   }
 
   private final class ESIndexerThread implements Runnable {
@@ -543,18 +537,12 @@ public final class IndexCollection {
             if (field.equalsIgnoreCase("id") || indexableFields.length == 1) {
               // Single value fields or "id" field
               Object value = document.getField(field).stringValue() != null ? document.getField(field).stringValue() : document.getField(field).numericValue();
-              switch (field) {
-                case "published_date":
-                  builder.timeField(field, value);
-                default:
-                  builder.field(field, value);
-              }
+              builder.field(field, value);
             } else {
               // Multi-valued fields
               Object[] values = Stream.of(indexableFields).map(f -> f.stringValue() != null ? f.stringValue() : f.numericValue()).toArray();
               builder.array(field, values);
             }
-
           }
 
           builder.endObject();
@@ -575,15 +563,12 @@ public final class IndexCollection {
         int skipped = segment.getSkippedCount();
         if (skipped > 0) {
           counters.skipped.addAndGet(skipped);
-          LOG.warn(input.getParent().getFileName().toString() + File.separator +
-              input.getFileName().toString() + ": " + skipped +
-              " docs skipped.");
+          LOG.warn(input.getParent().getFileName().toString() + File.separator + input.getFileName().toString() + ": " + skipped + " docs skipped.");
         }
 
         if (segment.getErrorStatus()) {
           counters.errors.incrementAndGet();
-          LOG.error(input.getParent().getFileName().toString() + File.separator +
-              input.getFileName().toString() + ": error iterating through segment.");
+          LOG.error(input.getParent().getFileName().toString() + File.separator + input.getFileName().toString() + ": error iterating through segment.");
         }
 
         segment.close();
@@ -734,9 +719,7 @@ public final class IndexCollection {
     public void destroyObject(PooledObject<SolrClient> pooled) throws Exception {
       pooled.getObject().close();
     }
-
   }
-
 
   private class ESClientFactory extends BasePooledObjectFactory<RestHighLevelClient> {
 
@@ -746,13 +729,8 @@ public final class IndexCollection {
       credentialsProvider.setCredentials(AuthScope.ANY, new UsernamePasswordCredentials(args.esUser, args.esPassword));
       return new RestHighLevelClient(
           RestClient.builder(new HttpHost(args.esHostname, args.esPort, "http"))
-              .setHttpClientConfigCallback(
-                  (HttpAsyncClientBuilder httpClientBuilder) ->
-                      httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider))
-              .setRequestConfigCallback(
-                  (RequestConfig.Builder requestConfigBuilder) -> requestConfigBuilder
-                      .setConnectTimeout(args.esConnectTimeout)
-                      .setSocketTimeout(args.esSocketTimeout))
+              .setHttpClientConfigCallback(builder -> builder.setDefaultCredentialsProvider(credentialsProvider))
+              .setRequestConfigCallback(builder -> builder.setConnectTimeout(args.esConnectTimeout).setSocketTimeout(args.esSocketTimeout))
       );
     }
 
@@ -766,7 +744,6 @@ public final class IndexCollection {
       pooled.getObject().close();
     }
   }
-
 
   public void run() throws IOException {
     final long start = System.nanoTime();

